File: Sorting/ShellSOrt/ShellSOrt.cpp
#include<iostream>
 
using namespace std;
 
// A function implementing Shell sort.
void ShellSort(int a[], int n)
{
	int i, j, k, temp;
	// Gap 'i' between index of the element to be compared, initially n/2.
	for(i = n/2; i > 0; i = i/2)
	{
		for(j = i; j < n; j++)
		{
			for(k = j-i; k >= 0; k = k-i)
			{
				// If value at higher index is greater, then break the loop.
				if(a[k+i] >= a[k])
				break;
				// Switch the values otherwise.
				else
				{
					temp = a[k];
					a[k] = a[k+i];
					a[k+i] = temp;
				}
			}
		}
	}
}
int main()
{	
	int n, i;
	cout<<"\nEnter the number of data element to be sorted: ";
	cin>>n;
 
	int arr[n];
	for(i = 0; i < n; i++)
	{
		cout<<"Enter element "<<i+1<<": ";
		cin>>arr[i];
	}
 
	ShellSort(arr, n);
 
	// Printing the sorted data.
	cout<<"\nSorted Data ";
	for (i = 0; i < n; i++)
		cout<<"->"<<arr[i];
 
	return 0;
}

File: Sorting/ShellSort/java/ShellSort.java
public class ShellSort{

    public static int[] shellSort(int arr[]){
        int temp;
        int n = arr.length;
        // Gap of size n/2 initially
        for(int i = n/2; i>0; i=i/2){
            for(int j = i; j<n; j++){
                for(int k=j-i;k>=0;k=k-i){
                    // break loop if is higher, if not switch
                    if(arr[k+i]>=arr[k]){
                        break;
                    }else{
                        temp = arr[k];
                        arr[k] = arr[k+i];
                        arr[k+i] = temp;
                    }
                }
            }
        }
        return arr;
    }


    public static void main(String[] args) {
        int array[] = {473, 78, 92, 5, 18, 7, 65, 89, 7};
        int sortedArray[];
        System.out.println("Before: ");
        int n = array.length;
        for (int i = 0; i < n; ++i) {
            System.out.print(array[i] + " ");
        }
        System.out.println();

        // shell sort
        sortedArray = shellSort(array);        
        System.out.println("After: ");
        n = sortedArray.length;
        for (int i = 0; i < n; ++i) {
            System.out.print(sortedArray[i] + " ");
        }
        System.out.println();
    }
}